returnn/tf/layers/basic.py

```diff
@@ -917,17 +946,33 @@ def get_out_data_from_opts(cls, name, sources=(), start=None, size=None, **kwarg
     :param int|None size:
     :rtype: Data
     """
-    start = start.output.copy_as_batch_major()
-    indices_data = start.copy_template(name="%s_gather_indices" % name)
-    indices_data = indices_data.copy_add_spatial_dim(spatial_dim_axis=start.batch_ndim, dim=size,
-                                                     auto_time_dim_axis=False)
+    from returnn.tf.util.data import DimensionTag
+    start_data = start.output.copy_as_batch_major()
+    input_data = sources[0].output.copy_as_batch_major()
+    input_t = input_data.placeholder
+    indices_data = start_data.copy_template(name="%s_gather_indices" % name)
+    if isinstance(size, int):
+      tag = DimensionTag(
+        kind=DimensionTag.Types.Spatial,
+        description="time_sliced",
+        batch=start_data.batch,
```

Comment:
It seems like the `batch=start_data.batch` argument leads to an error since a recent commit in `Data._adapt_batch_consistent_dim_tags`:

```
  File "/mnt/projects/i6/returnn/returnn/tf/util/data.py", line 1393, in Data.__init__
    line: self._adapt_batch_consistent_dim_tags()
    locals:
      self = <local> Data{'segments_gather_indices', [B,F|'time_sliced'(3)], dtype='int32', ctx=loop('time'[B])}
      self._adapt_batch_consistent_dim_tags = <local> <bound method Data._adapt_batch_consistent_dim_tags of Data{'segments_gather_indices', [B,F|'time_sliced'(3)], dtype='int32', ctx=loop('time'[B])}>
  File "/mnt/projects/i6/returnn/returnn/tf/util/data.py", line 1689, in Data._adapt_batch_consistent_dim_tags
    line: self._dim_tags = tuple(
            tag.get_for_batch_ctx(batch=self.batch, ctx=self.control_flow_ctx) for tag in self._dim_tags)
    locals:
      self = <local> Data{'segments_gather_indices', [B,F|'time_sliced'(3)], dtype='int32', ctx=loop('time'[B])}
      self._dim_tags = <local> (DimensionTag{B}, DimensionTag{'time_sliced'(3)})
      tuple = <builtin> <class 'tuple'>
      tag = <not found>
      tag.get_for_batch_ctx = <not found>
      batch = <not found>
      self.batch = <local> BatchInfo{B}
      ctx = <not found>
      self.control_flow_ctx = <local> ControlFlowContext{loop('time'[B])}
  File "/mnt/projects/i6/returnn/returnn/tf/util/data.py", line 1690, in <genexpr>
    line: self._dim_tags = tuple(
            tag.get_for_batch_ctx(batch=self.batch, ctx=self.control_flow_ctx) for tag in self._dim_tags)
    locals:
      self = <local> Data{'segments_gather_indices', [B,F|'time_sliced'(3)], dtype='int32', ctx=loop('time'[B])}
      self._dim_tags = <local> (DimensionTag{B}, DimensionTag{'time_sliced'(3)})
      tuple = <builtin> <class 'tuple'>
      tag = <local> DimensionTag{'time_sliced'(3)}
      tag.get_for_batch_ctx = <local> <bound method DimensionTag.get_for_batch_ctx of DimensionTag{'time_sliced'(3)}>
      batch = <not found>
      self.batch = <local> BatchInfo{B}
      ctx = <not found>
      self.control_flow_ctx = <local> ControlFlowContext{loop('time'[B])}
  File "/mnt/projects/i6/returnn/returnn/tf/util/data.py", line 131, in DimensionTag.get_for_batch_ctx
    line: assert not self.batch
    locals:
      self = <local> DimensionTag{'time_sliced'(3)}
      self.batch = <local> BatchInfo{B}
```

It seems like, if the dimension is static (so if `size` is an int in our case), the `batch` information should not be set. If I just remove the `batch` argument, it works. Also, the error only happens for the case of `size` being static and not if `size` is set to `None` and is therefore dynamic. Should I just remove the `batch` argument for the static case?

Reply:
Yea, remove it for the static case.